program-runtime/src/loaded_programs.rs

```diff
@@ -382,6 +382,19 @@ impl LoadedProgram {
     }
 
     pub fn to_unloaded(&self) -> Option<Self> {
+        #[cfg(test)]
+        if !matches!(&self.program, LoadedProgramType::TestLoaded(_)) {
+            return None;
+        }
+        #[cfg(not(test))]
+        if !matches!(
+            &self.program,
+            LoadedProgramType::LegacyV0(_)
+                | LoadedProgramType::LegacyV1(_)
+                | LoadedProgramType::Typed(_)
+        ) {
```

Comment:
is this a valid thing to do, or should we be logging something here?

Reply:
why not log where the error occurs? should this method be reused, now we have to remember to duplicate the logging. maintenance complexity that we don't need

Reply:
> We already do:
> 
> https://github.com/solana-labs/solana/blob/174023ef6dc8cbc9303a0af496f51720cab0f1c6/program-runtime/src/loaded_programs.rs#L1098

this is the thing I wasted hours on while replaying the ledger the other day. `unload_program` is only used in tests. Please let's add a comment that says so and make it `#[cfg(test)]`. 

The real thing is `unload_program_entry`, and we don't need an error log there. `to_unloaded` was originally designed to return `None` to signify "this program type can't be unloaded", see https://github.com/Lichtso/solana/blob/174023ef6dc8cbc9303a0af496f51720cab0f1c6/program-runtime/src/loaded_programs.rs#L1125

Reply:
The method is not even used in tests, it is not used at all.